### ISSUE TITLE
I run all tests in a order

### ISSUE THREAD

**Issue by cristopher-rodrigues** [cypress-io/cypress#390]
How can I run all tests in a order without rename the files using a number to determine the order to execute?

There is a another way that can I run the tests in a custom order?

For now, I’m renaming the files using the prefix 1_xxx.js, 2_xxx.js

**Comment by brian-mann** [comment]
No, you shouldn't need to run tests in any specific order. That is usually indicative of a testing anti pattern, whereby you are relying on state being built up between all the tests.

Can you explain your situation / use case?

**Comment by cristopher-rodrigues** [comment]
For example: before my other tests, i need create a user. But what i really want is  https://docs.cypress.io/docs/configuration#section-global

```
// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your other test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/guides/configuration#section-global
// ***********************************************************

// Import commands.js and defaults.js
// using ES2015 syntax:
import "./commands"
import "./defaults"

// Alternatively you can use CommonJS syntax:
// require("./commands")
// require("./defaults")
```

**Comment by fvanwijk** [comment]
Ordering tests could be convenient if you want to run the often failing tests first. It saves you some time in CI.
Another use case is when you have a lot of scenarios and you want to organize them in a way to makes more sense than alphabetically.

**Comment by digitaldavenyc** [comment]
@brian-mann I have a use case for why it is really important to be able to run tests in order. We are using Cypress for functional live tests. 

When running live tests we are dealing with real customer data that is somewhat dependent on state, in order for us to run our tests we need to hit real endpoints and perform actual CRUD operations. If we cannot control what test packages are called in an order the tests will fail since clearing out data is a big part of live testing.

I believe this is somewhat related to this issue: https://github.com/cypress-io/cypress/issues/263

**Comment by digitaldavenyc** [comment]
Right now we are resorting to labelling our test folders `01_Users`, `02_Invoices`, etc... to control the test ordering which just makes me very sad.

**Comment by charlestonsoftware** [comment]
As @digitaldavenyc points out full blown web apps often test complex system interactions.    

Should every test be able to run independently and in a vacuum?  Sure.   In theory.   Real world and theoretical science do not always get along. It may be anti-pattern but coloring 100% within "the pattern" lines means my test scripts take 27 hours to complete instead of 27 minutes.


Use Case--

My app allows users to add locations to a data set on the back end.  On the front end it renders a map of locations.   

Front-End test: Make sure 10 locations are returned from Charleston SC

This is heavily dependent on the data set that includes two primary elements: location data + user-set options such as "default radius to be searched".

I could write the test to first pull the options data (default: 5 mile radius), THEN pull all location data (5k entries), THEN let the script loop through all 5,000 to calculate the distance to create the  baseline of valid results, THEN run the front-end query and make sure it returns the same list my test calculated.

I've now created another point of failure as my test script code is more complex and prone to errors.  Not too mention it takes a LOT longer to run by not being able to make data assumptions.

OR 

I could write a test that ASSUMES my "load_5k_locations_spec.js" has executed and passed.  The test is now  pull the option data, load a  "valid_5_miles_from_charleston.json" fixture, run the front-end query and compare it against the displayed location divs.   

An order of magnitude faster AND far less complex test script code.


NOW... take the above and run the test for five different radius options.   

I'd rather pull a valid_5_mile.json, valid_10_mile.json, etc. and compare against an assumed set of data that can ONLY be a valid assumption if I am certain my "load_5k_locations_spec.js" ran BEFORE all my "5miles_from_charleston.js" and "10miles_from_charleston.js" scripts ran.


Bonus Points: Have a Cy.passed( "load_5k_locations_spec" ) command the returns true|false if the specified test passed on the prior run -- makes it easy to completely skip a test if a prior test run failed.



No - not perfect "by the pattern" rulebook implementations, but in the real world people have deadlines.    Making tools that are malleable enough to meet user's needs versus doing things "strictly by the book" is what makes them powerful.    

I'm fairly certain chain saws are not designed to be used to make ice sculptures - but people do it.  If you try to do the same and cut your arm off because you don't have the skills to use that tool in a manner it was not intended, consider it a "learning curve".

**Comment by bahmutov** [comment]
I strongly suggest NOT going down the path "real world requires running spec files in order". It is just JavaScript and you can easily set state before each test the way you want, without relying on the previous test. In fact we are working on automatic test balancing so different specs will be running on different CI machines and in different order (if you use `--parallel` flag). So any test order will not be guaranteed.

Instead, split tests into functions and just call these functions to set everything up.

**Comment by surfjedi** [comment]
I agree with @cristopher-rodrigues  that running tests in order has major advantages - speed wise.  You can save time but not reloading the login page(or doing requests for i) each time

**Comment by RadhikaVytla** [comment]
How to run login spec before create user spec?

**Comment by araymer** [comment]
It is, indeed, a huge pain to run e2e tests with an SSO redirect (which can't be removed) if you cannot make sure your login procedure runs as the first test to set a valid token. Otherwise, you have to write logic in every single test file to handle that, which seems silly.

**Comment by jameseg** [comment]
@RadhikaVytla you could run the login_spec first as an isolated job (CI), and depend all other test specs on that job being successful.  Could save you some time if login isn't working.

**Comment by jameseg** [comment]
@araymer can you set the token through an auth api call?  If so, just use a beforeEach (write it once) and this way each test can use it's own separate auth token

**Comment by jennifer-shehane** [comment]
@araymer You can also move this to the support file within a `beforeEach` as described here: https://on.cypress.io/writing-and-organizing-tests#Support-file

**Comment by spcbrn** [comment]
The truth is that there are so many variable factors between different applications that it's a bit silly to just assert that one pattern 'should' work in any environment or in the context of any system, lest you've done it wrong.  Our testing flow relies heavily on carefully sequenced actions just by it's nature as an application and what it does, and there's no way around that without failing to actually test the UI fully e2e.

Test sequencing and sequential interdependency is ABSOLUTELY necessary in some applications, as is being expressed by many others here.  Cypress is amazing and powerful (thank you so much) and we've scripted our way around this issue for now, but of course this feature would be powerful and appreciated eventually :).

**Comment by lazurey** [comment]
Being able to specify the order could be very useful when testing a process composed by several steps. I'd like each step to be a standalone test, but they need to be executed in a specified order. Currently I can only follow the advice that naming them with `01`, `02`..

**Comment by id-dan** [comment]
Hi, All! 
I have the same problem. I want to run tests in a specific order: 1.SigUp, 2.Login, 3.ShareTool, 4.Publication and then 5.Account. However tests run like on this screenshot:
![image](https://user-images.githubusercontent.com/52101871/59999337-682d2680-966a-11e9-906a-3fec77ea795c.png)

**Comment by loxator** [comment]
Just found an easy hack, add a number before your file or folder and it will run those in that order. 

For example, @id-dan you can do something like 01-SignUp/01-SignUp.feature/, 01-SignUp/02-Login.feature/.....05-Account/... and so on.

Cypress should pick up the files in that order and run them.

**Comment by Julyanajs** [comment]
Thx @loxator your suggestion solved my problem! =)

**Comment by malimccalla** [comment]
This would be a great feature. I currently have two test files that take a lot longer to run and are "more flakey" than others resulting in the occasionally fail for non legitimate reasons.

Due to the file names (`upgrade_page.test.ts` & `settings_page.test.ts`) both these files run at the end of a 15+min e2e suite. I could save time and money with my CI provider if I were able to run these tests first so I can bail out of remaining tests if they fail

**Comment by fusionshoes** [comment]
Any news on this @brian-mann ? Cy best practice for each test to run standalone is excellent and perfectly architected, but there's definitely an added convenience to being able to order tests, and I'm curious if you and your guru coders are considering it?
Thanks!

**Comment by majackson** [comment]
To reiterate, this feature would be really useful in order to fun often-failing/fragile tests first, thereby ensuring that failures are flagged as quickly as possible. Some workarounds for this have been described in this issue thread (such as numbering tests in intended order), but this isn't really a solution, as it relies on a cumbersome manual renaming of test files. If test ordering could be specified by some configuration file, this file could be generated and updated by subsequent test runs, putting frequently-failing tests first.

I would love to get an update on whether this feature is being considered.

**Comment by jholland918** [comment]
> No, you shouldn't need to run tests in any specific order. That is usually indicative of a testing anti pattern

That's correct for **unit** tests, however integration and end to end tests are a different beast. I appreciate the additional power cypress gives us for injecting state directly into the page and the cy.request feature to avoid having to do things like ordered tests. But sometimes you just need ordered tests and the whole "tests should run in any order" is simply a cargo cult mindset without considering the context.

**Comment by dwelle** [comment]
I don't think "tests should be able to run in any order" is a cargo cult. Even in e2e, randomizing test order (which cypress doesn't support yet, though) can reveal bugs in DB or anywhere where state is preserved.

That being said, I too would like this feature implemented simply from the standpoint of debugging failing tests, as many have mentioned before, and not just in CI. For example, the issue [I've posted earlier](https://github.com/cypress-io/cypress/issues/5475) could be reproduced only by running two spec files in consecutive order, and to isolate this I had to manually remove all other spec files so that only the two spec failing files remained, and I could iterate much more quickly in debugging (instead of waiting 30 minutes every time until previous spec files finished).

**Comment by bahmutov** [comment]
@dwelle I think you better use `testFiles` or `--spec` to select the spec files you want to run on CI, no need to remove others. Also pick the tests to run in a spec file using `it.only`.

**Comment by dwelle** [comment]
@bahmutov yes, I'm doing that. But you can't use `--spec` in headed mode, nor does `describe.only` do anything when you hit the `run all tests` button in cypress runner.

**Comment by bahmutov** [comment]
You can use testFiles option in headed mode

**Comment by dwelle** [comment]
@bahmutov yea, sorry --- didn't know about that option and I don't know why I skipped over that when reading your reply. WTBS, from what it looks like the test-order will still be name dependent (but I'd have to check that).

**Comment by ginna-baker** [comment]
I want to run my most basic tests first, so that if they fail, I won't need to bother running the rest of the tests.  It makes my CI fail faster and lets me start debugging earlier.

**Comment by bahmutov** [comment]
@ginna-baker you can do first "sanity" run with Cypress before running all tests, something like this
```shell
npx cypress run --spec cypress/integration/sanity-spec.js
npx cypress run
```

**Comment by vctormb** [comment]
I think I'm going to rename all my .spec files with a number as a prefix, because I have a scenario that it's important to run first than other tests.

**Comment by chloeong** [comment]
testFiles config is work for me!

**Comment by jeanlescure** [comment]
A more detailed example of the config @chloeong refers to can be found here: https://stackoverflow.com/a/59690611/2731075

**Comment by craig-dae** [comment]
I'm looking into this right now. The reason is: 
- We have certain data facets in our app. Things like countries, teams, etc. 
- Part of our requirements is that these are keyed by guids, not by counters. This means that each time our dev db gets rebuilt, all the ids change
- Creating test fixtures becomes impossible, since we need to save properties of objects using guids
- I want to start my tests by first, calling all the endpoints for things like countries, teams, etc, and using it to make fixtures, so I can get the IDs of things by name
- Then I want to use fixtures I've made for my other objects, and use the fixtures for the facets to do the ID lookups. 

All that stuff needs to run first so my fixtures are all set up with proper guids and stuff. I'm doing this in GitHub Actions. 

I think @bahmutov has the right idea. I would take a spec and drop it in a folder like `cypress/setup/create-fixtures.spec.ts`. Then I'd first run cypress against this one spec, setting everything up that I need for the rest of the tests. 

Next, I'd run the rest of my tests in parallel. 

This is my plan. I haven't implemented it yet. I might start tomorrow.

**Comment by samgre** [comment]
Our use case here would be to first run tests starting with an empty environment to verify our user onboarding process works correctly.

After user onboarding related features have been checked, I would like to run rest of the suite against data created by the onboarding tests.

**Comment by sktguha** [comment]
Hi any progress on this, I believe it will be a great feature. also what we can do is still parallel testing by maybe specifying which test depends on which. But still, independent groups of tests can run parallelly. @brian-mann

**Comment by sktguha** [comment]
> I'm looking into this right now. The reason is:
> 
> * We have certain data facets in our app. Things like countries, teams, etc.
> * Part of our requirements is that these are keyed by guids, not by counters. This means that each time our dev db gets rebuilt, all the ids change
> * Creating test fixtures becomes impossible, since we need to save properties of objects using guids
> * I want to start my tests by first, calling all the endpoints for things like countries, teams, etc, and using it to make fixtures, so I can get the IDs of things by name
> * Then I want to use fixtures I've made for my other objects, and use the fixtures for the facets to do the ID lookups.
> 
> All that stuff needs to run first so my fixtures are all set up with proper guids and stuff. I'm doing this in GitHub Actions.
> 
> I think @bahmutov has the right idea. I would take a spec and drop it in a folder like `cypress/setup/create-fixtures.spec.ts`. Then I'd first run cypress against this one spec, setting everything up that I need for the rest of the tests.
> 
> Next, I'd run the rest of my tests in parallel.
> 
> This is my plan. Has anyone done anything like this?

@craig-dae have you managed to pull off what you mentioned above ? If so , do you have any tips etc to share regd this ? As I am also facing a similar problem where 4 tests depend on 1 test executed first

**Comment by craig-dae** [comment]
@sktguha Nope. In my case, the test-users shouldn't be re-created often, if at all. I had to recreate them when a new developer ran a script they weren't supposed to run, so I had to run my scripts over from the beginning to create test users. 

My solution is that, if you re-create test users, you have to sign the eula immediately. For the eula cypress test, the test takes them directly to the eula page, rather than relying on the app to redirect them to the eula page.

In a way, this is more of an issue with Cognito than with Cypress. Cypress is trying to enforce a very opinionated testing philosophy. Cognito should provide a "dev mode", where creating and modifying users for the purpose of tests is trivial.

**Comment by Wenish** [comment]
Sorry for the little rant ;)
First of all I really love cypress and its a mile a head of everything i saw before.
But its sad to see that for real world problems most of the times i just see the answer "This isnt the correct way todo" 
i really wish the maintainers off this project would open up their ears and minds and help to solve those problems.
IMO an application should be built so it works for a certain e2e testing tool. the e2e testing tool should be built that it works with real world applications.

**Comment by craig-dae** [comment]
@Wenish I'm on both sides of the fence here. 

I actually prefer frameworks that are very opinionated about the right way things should be done. They tend to be VERY good at doing those things. If they try to accommodate every use-case, including use-cases that are incorrect, they become checkbox heroes like Selenium. They become average at a lot of things, and they have to support all these features that aren't at the core of their mission. Instead, Cypress is very very good, because it refuses to burn cycles on helping people test the wrong way. 

That said, for this issue, I would hope that they'd give me their opinionated solution, rather than say "newp". I get why they don't want to allow us to order tests: because it will lead to bad tests that are not stateless. But, what is the right solution for things like my EULA problem?

**Comment by Wenish** [comment]
@craig-dae the second part you wrote is what i hope. okey that the tools dosnt do it out of the box. but help us get a good workaround going. (the other topic i have is the SSO with azure ad and this is  imo a must have. so many entreprises use SSO with azure ad. and tickets are open for years and there is still no okey solution)

**Comment by fusionshoes** [comment]
Jumping in quickly...
I agree with @craig-dae - cy has a very solid laser focus that I very much appreciate, and it blows everything else out of the water for e2e. And it's getting better with each passing week, so.... thank you, Cy teams.

For ordering tests specifically ("Ability to run spec files in a specific order"), I simply use:
```
  "testFiles": [
    "folder1/test1.spec.js",
    "folder2/test1.spec.js",
    "folder2/test2.spec.js",
   ],
```
I hope that helps someone who might need that...
Aloha

**Comment by craig-dae** [comment]
@fusionshoes this works great if you only test local. 

Our app has grown to over 300 tests, and it runs in cicd in GitHub Actions using 6 different boxes for tests. When you run tests this way, they are run at random to ensure isolation, and naming your tests does not affect the order. 

The only order that is not entirely random in cicd is that, new tests are put at the front. The idea is that they are the least-likely to pass, and that you want to know as quickly as possible.

**Comment by fusionshoes** [comment]
Sounds like a nice setup, @craig-dae . We do run these ci-cd (circle), using paralellism (which does introduce an element of randomness, since no flags are currently being passed - such as test timings - within that call).

**Comment by Threnos** [comment]
In my use case it would be greatly beneficial to have order.

Let's imagine, I want to test that a user can add a department to location. Preparations for this test would include user and location creation, and I surely want to test these first. But alphabetically "department" stands before "user" and "location".

If location creation fails, then there is no point in even trying to test departments.

**Comment by Threnos** [comment]
> It is just JavaScript and you can easily set state before each test the way you want, without relying on the previous test

What about integration and e2e testing, as what Cypress claims to be used for?

I want to test full-blown application that is made up of several services and 2 databases.
Very simple case: a user must be able to assign product to category. Preparations for this test would include creation of user, product and category. I surely want to test these first.
If any of entities creation fails, then there is no point in even trying to test assignment.

**Comment by craig-dae** [comment]
@bahmutov depends on the app. Also, Cypress tends to be run by the frontend guys, and it's not always easy to get the backend guys to do things the right way. 

I have one app where it is very easy to set things up making cy.request calls in the test setup. I have another app where the data is so convoluted, coming from a combination of dynamo and elastic search, that it is difficult to set it up using the web services I have. Our app literally emulates other AWS regions, and requires you to manually create a VPC peering connection as part of the setup for some of the data. I can't use javascript to emulate creating an AWS account, and VPC-peering it to our VPC, so that the webapp will show "VPC-peered successfully!". I also can't use Javascript to create Cognito users on the fly. 

There are some limitations that we struggle with. We try to use Cypress's ability to intercept web service calls with mock data to get around these limitations. But that means that our tests are pure frontend, and not integration tests, in those cases.

**Comment by jennifer-shehane** [comment]
We have not closed this issue as `wontfix`, so it is still under consideration as a proposed feature. We do read the comments and are always willing to reconsider our opinions on the best way to run tests once more use cases come to light. 

As noted above, there is a complication introducing test file ordering combined with Cypress parallelization. We have our own algorithm that runs the longest running spec files firsts in order to spread out the load to get the shortest runtime across machines (with new specfiles running very first since we have no previous data to know how long it will take to run). So combining any type of ordering with the algorithm we use with our `--parallel` flag would have to be considered which makes this interesting to implement. 

We would also have to come up with a new API to to define ordering of files. 

If you have some specific specs that need to run first before others. I would consider using our [`--spec`](https://on.cypress.io/command-line#cypress-run-spec-lt-spec-gt) flag. So, in your CI having a job that runs first with `cypress run --spec 'setup_db.js'` for example. Then after this job is complete, trigger a new job that runs `cypress run` with all specs. This may not work in all situations, but it's a way to run a single / multiple files in an order and then also run the rest of the files in parallel later to save time.

We have limited resources, so we have to make hard decisions on what we work on. This issue is still in the 'proposal' stage, which means no work has been done on this issue as of today, so we don't have an estimate on when this will be delivered.

**Comment by craig-dae** [comment]
@jennifer-shehane I have tried that. However, it has a problem that maybe you can help solve. 

IIRC, you can't run a spec file that is not contained in `integration`. I forget the error, but it didn't seem to work. 

This means that if I explicitly run spec files in phase 1 that are contained in `integration`, they will be re-run in phase 2. I don't want them to be re-run. Is there a way to mark a test as "do-not-run" when you run all tests?

Also, a "do-not-run" option would be great for another reason. We have a very large project that had a part of it put on hold, so we could focus on other parts. This leaves a section of code and tests that we don't want to delete, but we also don't want to be forced to continue to maintain. I would like the tests left in `integration`, but I have no desire to fix them if they start failing, and I don't want to run them. But if we come back to this part of our app, we would like to begin using them again. Our current strategy is to just move all the tests to an `archive` folder outside of `integration`.

**Comment by bahmutov** [comment]
Could you put .skip on those tests like describe.skip and it.skip?

**Comment by craig-dae** [comment]
@bahmutov didn't even realize that existed. I see it now. Yes, I'll add that. Thanks!

**Comment by ddehart** [comment]
I stumbled upon this feature request searching for a solution for an adjacent use case.

I have tests that are heavily dependent upon an API responding in ways that I expect. I'd like to be able to write a test that's sort of like a pre-flight checklist using `cy.intercept` that asserts that the API responds as I expect before running any other tests and then bails on running the entire suite if that test doesn't pass.

I think this is also related to https://github.com/cypress-io/cypress/issues/518, so I'd need both the ability to specify that the pre-flight check runs first and that the entire run aborts if it fails.

**Comment by craig-dae** [comment]
@ddehart You know, that's a great point. Just today, we had a meeting about using Lighthouse to test our app for accessibility. There is a lighthouse/cypress plugin I was interested in using. If I go this route, I want my lighthouse tests to be completely separate from my integration testing, but I still want to use cypress to do it. 

I would like a way of running integration tests first, followed by lighthouse tests. 

A way of doing this might be to follow the Ansible pattern. In Ansible, you can apply tags at the playbook or task level. I could tag all my integration tests as `integration`, and lighthouse tests as `lighthouse`, and then do something like `cypress run --tags=lighthouse`, which would run everything with the lighthouse tag and ignore everything else. I could also have a tag called `always` which will always run, regardless of what tags I choose. 

I think that would work smoothly with cypress.

**Comment by bahmutov** [comment]
@craig-dae I would place different types of tests in different subfolders and then run them using `--spec ...` argument. For example if you want to run regular UI tests first, then run the Lighthouse tests

```
- npx cypress run --spec 'cypress/integration/ui/**/*.js`
- npx cypress run --spec 'cypress/integration/lighthouse/**/*.js`
```

Note: there is also https://github.com/bahmutov/cypress-select-tests but it relies on rewriting tests and in general I would not rely on it

**Comment by craig-dae** [comment]
@bahmutov that might actually do it for me. That looks like it will work in github actions. I think that would also solve @ddehart's problem. Thanks!

**Comment by jwetter** [comment]
Just a little piece of history as it relates to fitting users needs.

Intel used to dictate to its' customers the products it provided.  They normally would fit 70-90% of the customers needs.  Intel executives figured that because they produced a superior product that it didn't matter if customers had to implement the other 10-30% to complete the customers needs.  Competitors decided to listen to customers to given them 95-100% of their needs since it was minor tweaks to fit their needs.  Intel market share plummeted forcing them to change.  They still haven't recovered the market loss.

The point I am trying to make is that while Cypress does a lot of things really well, there still is a need from the users to have some structured order to certain testing scenarios.  For instance there are some tests that I have that need to be ran in order (some elements of integration where I check record creation and modification) but most of mine I would want ran in parallel.  To be against this idea of having some test execution in a specific order is to also acknowledge that while Cypress is a great design that it is not designed for long term usage because it is inflexible to the needs of the users.  Relying on plugins is not a good route as sooner or later they cease being maintained which means loss of efficiency in test execution or worse, incompatibility due to structure changes that are likely to occur in Cypress.

**Comment by craig-dae** [comment]
@jwetter I mean, while that all makes sense, @bahmutov literally just gave you a solution in the previous comment for your outlier scenario. 

Cypress is opinionated about how testing should work, rather than being checkbox heroes. This is why I like them. Their opinions have frustrated me, mostly because they have constantly ended up being correct. They've steered me to far better testing practices than I'd be engaging in if I were able to force them to bend to my incorrect ways of doing things.

But again, your use-case sounds like it is solvable pretty trivially, by organizing the phases of your tests by top-level directories in `cypress/integration`, and then running them using the command level wildcard examples above. The only problem I can think of with it is how it might behave in combination with the Dashboard?

**Comment by bahmutov** [comment]
@craig-dae and @jwetter using the wildcard pattern and Dashboard also works - just use the `group` parameter to put separate runs into the same logical recorded run
```
- npx cypress run --spec 'cypress/integration/ui/**/*.js` --record --group ui
- npx cypress run --spec 'cypress/integration/lighthouse/**/*.js` --record --group lighthouse
```

**Comment by jwetter** [comment]
@craig-dae and @bahmutov 
I have already implemented the workarounds listed here to suite my needs.  The steps taken are still just workarounds and while currently working for my needs, the cautionary story I point out in my post remains.

To be clear, I only posted because I think this tool has great promise to become an industry standard.  The reason Selenium is still an industry standard is because they changed their software to meet the needs of their users where reasonable.  Cypress could beat Selenium but only if it is meeting the needs of the users.

**Comment by jeanlescure** [comment]
Having this issue surpassing the **4 year** mark and given that users keep providing sensible use cases, it would be all-around beneficial to begin solving the issue by formalizing the `testFiles` option as the official way of running tests in order.

The following tasks would accomplish the aforementioned goal:

- Acknowledge that being able to use the `testFiles` option to run tests in order is a side-effect of [this line of code](https://github.com/cypress-io/cypress/blob/fec14520a4baef1357bd6fcc7a20f96157cf35df/packages/server/lib/util/specs.js#L156-L159) as well as [this other one right below the previous](https://github.com/cypress-io/cypress/blob/fec14520a4baef1357bd6fcc7a20f96157cf35df/packages/server/lib/util/specs.js#L176-L177), by writing code comments above each of these lines. The comments should specify that the side-effect is welcome and should not be tampered with (for example adding a `.sort()` in either of those would introduce an inevitably painful backwards compatibility problem to 4+ years worth of users that have implemented the `testFiles` option to run tests in order)
- Update the description on [this schema](https://github.com/cypress-io/cypress/blob/fec14520a4baef1357bd6fcc7a20f96157cf35df/cli/schema/cypress.schema.json#L45-L52) to mention that this is the official way of running tests in order
- Add a spec [here](https://github.com/cypress-io/cypress/blob/fec14520a4baef1357bd6fcc7a20f96157cf35df/packages/server/test/unit/config_spec.js#L473-L498) that verifies that when given `this.setup({ testfiles: ['./b.js', './a.js', '**/23/*.jsx', '**/15/*.jsx'] })` then checking `this.config` should contain `['./b.js', './a.js', '**/23/*.jsx', '**/15/*.jsx']` on the `testFiles` property (as in expecting the array order doesn't change)
- Finally update the testFiles entry in official [documentation](https://github.com/cypress-io/cypress-documentation/blob/master/source/guides/references/configuration.md) to formalize it as the official way of running tests in order

P.D. I originally had the intention of doing the tasks myself and submit a PR but I haven't had the time, so by submitting this reply do not think I expect the Cypress team to complete the tasks, it is truly an invite to any colleague and/or hobbyist to take the lead in making sure that the "unofficial" fix (that has earned me 200+ points [in Stackoverflow](https://stackoverflow.com/a/59690611/2731075) so far 😄 ) becomes a properly recognized feature :rocket:

**Comment by diggabyte** [comment]
Not all ordered tests are anti-pattern. We have a very valid use case.

We distribute and balances specs across workers. We use run duration and passing status to update the balancing after each pass. The value in doing so is that we place **failing tests first** on subsequent runs for **faster feedback for developers** when running under CI. 

Very often tests that pass locally might fail in CI. So devs might have to wait for many other specs to run before receiving feedback on a test they are trying to fix. This is a huge waste of time.

Run order should be left up to us. If someone wants to run sorted by alpha, simply pipe the files to `sort` on the command line. Or at least give us a flag to force cypress to respect ordering

**Comment by craig-dae** [comment]
I very much would like to echo @diggabyte's point that, any tests that have failed or have been flaky recently should be run first, followed by the order Cypress currently does (longest first I think?)

Cypress is great. This causes me to use it more. The more I use it, the more I rely on it. The longer my tests get. The more parallel boxes I run it on (8 now). The more impatiently I wait to see if my tests pass. The more I beat my head against my keyboard when a frequently-failing test gets run on the bottom half of my list of tests.


If Cypress just did this by default (flaky and failing tests first), @diggabyte would not have to manually order his tests (for the reason he stated). Seems like a relatively cheap LOE to add a lot of value.

**Comment by NPC** [comment]
Is there currently a way to define that if a test fails, then the whole run should fail? It could be a good addition to ordering — I'd place general (bird-eye-view) tests first, and if they passed — move on to more detailed ones. 

Currently our full set takes about an hour, which basically means that if something fails — we'll only know about it the next day (still better than hearing about it from our end, so I'm thankful for what Cypress allows us to do). Ability to order tests AND configure cypress to stop on first error (or specify which tests are “critical” — there may already be a way to do it?) would help improve the responsiveness of this process.

**Comment by MuckT** [comment]
@NPC checkout [cypress-fast-fail](https://www.npmjs.com/package/cypress-fail-fast); however, sometimes it doesn't play nice with other plugins.

Tests can be ordered in folders like this:
```
test-folder
  test.00.spec.js
  test.01.spec.js
  test.02.spec.js
```

**Comment by AlliterativeAlice** [comment]
I would also appreciate being able to do this without having to resort to the workaround of changing my test file names, so that I can run the tests that fail most often first.

**Comment by impurist** [comment]
Been wanting a `determinate test order escape hatch` for several years now. 
While I appreciate Cypress have a testing philosophy. That philosophy does not fit the requirements of many users. 
It is `Ivory tower` thinking. And it's a huge. black mark against Cypress.  
Forcing us to use poor naming convention work arounds when a  simple config option allowing a determinate order is all we need. 
 Escape Hatches are Required

**Comment by bahmutov** [comment]
For Cypress v10 just list the specs in the order you want them to run 
```js
const { defineConfig } = require('cypress')

module.exports = defineConfig({
  e2e: {
    // baseUrl, etc
    supportFile: false,
    fixturesFolder: false,
    setupNodeEvents(on, config) {
      config.specPattern = [
        'cypress/e2e/spec2.cy.js',
        'cypress/e2e/spec3.cy.js',
        'cypress/e2e/spec1.cy.js',
      ]
      return config
    },
  },
})
```
Of course, if you want to parallelize them using Cypress Dashboard it will change the order based on timings / failed tests first / new specs first.

**Comment by djorgji** [comment]
> Of course, if you want to parallelize them using Cypress Dashboard it will change the order based on timings / failed tests first / new specs first.

Is there a way to turn this off, or to tell it that or to add some spec dependency? We have a Registration test we want to run first before running any other test or they will all fail?

**Comment by gowtham-ncompass** [comment]
> For Cypress v10 just list the specs in the order you want them to run
> 
> ```js
> const { defineConfig } = require('cypress')
> 
> module.exports = defineConfig({
>   e2e: {
>     // baseUrl, etc
>     supportFile: false,
>     fixturesFolder: false,
>     setupNodeEvents(on, config) {
>       config.specPattern = [
>         'cypress/e2e/spec2.cy.js',
>         'cypress/e2e/spec3.cy.js',
>         'cypress/e2e/spec1.cy.js',
>       ]
>       return config
>     },
>   },
> })
> ```
> 
> Of course, if you want to parallelize them using Cypress Dashboard it will change the order based on timings / failed tests first / new specs first.

Is there a way where i can run particular spec file twice ?
```
`const { defineConfig } = require('cypress')

module.exports = defineConfig({
  e2e: {
    // baseUrl, etc
    supportFile: false,
    fixturesFolder: false,
    setupNodeEvents(on, config) {
      config.specPattern = [
       'cypress/e2e/spec1.cy.js',
        'cypress/e2e/spec2.cy.js',
        'cypress/e2e/spec3.cy.js',
        'cypress/e2e/spec1.cy.js',
      ]
      return config
    },
  },
})`

```
For the above config , spec1.cy.js is running only once. Is there any way to run the same file twice or more?

**Comment by skiKrumbRob** [comment]
Has there been any update on this for an official test ordering solution since last conversation over a year ago? We have a need for this on our project as well. @jennifer-shehane is it still CY's official stance that this will not be implemented?

**Comment by bahmutov** [comment]
@skiKrumbRob why is the solution with explicit spec order not working for you? If you need to run specs in specific order in parallel and have concrete requirements and example, open an issue in https://github.com/bahmutov/cypress-split and it would be simple to implement

**Comment by skiKrumbRob** [comment]
@bahmutov Forgive my learning brain here, but I digested this full thread yesterday trying to figure out if/how ordering can be done and perhaps my eyes glazed over that bit by the end of this monster thread. Pretty new to Cypress and learning on the fly. Is there any documentation for the explicit spec order that I could look through to get a better grip on where to set it up etc.?
And thanks for responding.

**Comment by kalcefer** [comment]
Hello, I have this folder structure.
![image](https://github.com/cypress-io/cypress/assets/63726271/a8ed9ab2-64a8-417b-8123-8cbc044c2ab3)

Inside the folders. the same structure of test suites

Inside the folder, the tests are performed in the correct numbering order.

01-02-03...10-11-12...20-21-22

But at full startup, the folders are always executed in a strange order

01-02-03-04-05-06-**08-10-11-12-13-07-09**

Inside folders 05, 07, 09 there are also folders with the same name structure.

all tests are independent of each other and the order is not affected.

but the tests differ in complexity and weight. And I would like to understand. where is my mistake.
